Iterators for shardedDB

winnie-byun opened the pull request:
## Proposed changes

There are 3 kinds of shardedDB
1. shardedDBChanIterator
   - This is created by calling `NewChanIterator` (this is a new function)
   - There are iterators and workers for each shardDB (If there are 4 shards in DB, 4  iterators and workers will be created)
   - If `resultCh` is given, all items are written to `resultCh`, unsorted.
   - If `resultCh` is not given, new channels are created for each DB. Items are written to corresponding channels in binary-alphabetical order. The channels can be gained by calling `Channels()`.
   - Each channels iterates item in sorted order.
2. shardedDBIterator
   - This is created by calling `NewIterator`, `NewIteratorWithStart`, `NewIteratorWithPrefix` (this is the original implementation of `Iteratee` interface)
   - contains `shardedDBChIterator`
   - There are 1 main channel and 1 worker that combines all shards.
   - The first `Next()` call could take some time. (Time for workers to fill channels)
3. shardedDBIteratorUnsorted
   - This is created by calling `NewIteratorUnsorted` (this is a new function)
   - The interface is same as `shardedDBIterator` except that it returns items unsorted and is much faster.
    
<img width="1920" alt="Screen Shot 2020-12-24 at 12 21 06 PM" src="https://user-images.githubusercontent.com/13597401/103057376-a51d0500-45e2-11eb-8837-523f6ac041c6.png">
<img width="961" alt="Screen Shot 2020-12-24 at 12 21 13 PM" src="https://user-images.githubusercontent.com/13597401/103057377-a64e3200-45e2-11eb-8f88-753da35b7cdd.png">

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues


## Further comments

aidan-kwon commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +153,188 @@ func (pdb *shardedDB) Close() {
 func (pdb *shardedDB) NewIterator() Iterator {
-	// TODO-Klaytn implement this later.
-	return nil
+	return pdb.newIterator(func(db Database) Iterator { return db.NewIterator() })
```
Could you rename `pdb` to `sdb` or `db`? `pdb` was originated from "partitionaedDB"

winnie-byun replied:
I renamed it to `db` and `it` for functions related to the iterator. I will make another PR to change the names in other functions.

aidan-kwon commented:
@winnie-byun I think we better define another interface for `shardedDBIterator` instead of using `iterator` or use the struct itself. You assumed `shardedDBIterator` doesn't iterate in alphabetical order, but `Iterator` has the characteristic iterating in binary-alphabetical order. Probably making `shardedDBIterator` to iterate in order could be one of the best options.

aidan-kwon commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +153,188 @@ func (pdb *shardedDB) Close() {
+}
+
+func (it *shardedDBIterator) Release() {
+	it.ctx.Done()
```
I cannot understand this line. As far as I know, it cannot send a message to `ctx.Done()` channel. Could you explain what does it for?

winnie-byun replied:
Thanks for pointing it out. It wasn't working properly.
I added cancel function to end context and test functions to test context.

winnie-byun commented:
@aidan-kwon I took your comment and changed `shardedDBIterator` to `shardedDBIteratorUnsorted`. I also added a new `shardedDBIterator` for alphabetic ordered iterator. 
So there are 3 iterators; `shardedDBIterator`, `shardedDBIteratorUnsorted` and `shardedDBChanIterator`.

ehnuje commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+// This is useful when you want to operate on each items in parallel.
+// If you want to get ordered items in serial, checkout shardedDB.NewIterator()
+// If you want to get unordered items in serial, checkout shardedDB.NewChanIteratorUnsorted().
+func (db *shardedDB) NewChanIterator(ctx context.Context, newIterator func(Database) Iterator) shardedDBChanIterator {
```
`NewChanIterator` seems to be used only in the database package, as it is wrapped by `newIterator`, is there any reason for making it public?

ehnuje replied:
Also, `Channels` and `Release` should be wrapped by the function of `shardedDBIterator` or `shardedDBIteratorUnSorted`, they should be private, shouldn't they?

winnie-byun replied:
I thought `shardedDBChanIterator` is needed more than `shardedDBIterator ` or `shardedDBIteratorUnSorted `.
If we are using `shardedDBIterator ` or `shardedDBIteratorUnSorted `, all data is carried with only one channel. If we use `shardedDBChanIterator`, data is carried with several channels and some operations could be done in parallel. I am considering using `shardedDBChanIterator` in DB migration for performance. I expect there could be other cases to use `shardedDBChanIterator` than `shardedDBIterator ` or `shardedDBIteratorUnSorted `.

ehnuje replied:
I understood. Thanks!

ehnuje commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+	return it
+}
+
+func (*shardedDBChanIterator) newChanWorker(it Iterator, resultCh chan entry, ctx context.Context) {
```
If its name starts with `new` prefix, I expect something to be returned but it does not.
How about change its prefix to `run`? 

It's a matter of taste, I think naming `Worker` here does not work well with the concept of worker since it does not receive any task from elsewhere. Just a comment, so do what you want.

ehnuje replied:
Also, you can push down the below codes into the `newChanWorker` function, which can reduce function parameters. 
```
		it.iterators[i] = newIterator(db.shards[i])
		it.resultChs[i] = make(chan entry, shardedDBSubChannelSize)
```

winnie-byun replied:
Function name
* I changed to the prefix to `run`.

Parameter
* To run `newIterator(db.shards[i])`, `newIterator` and the shard are needed. If the codes are run in `newChanWorker`, `newIterator` and the shard should be passed trough function parameters. I think the number of function parameters will be the same. Did I misunderstood what you are saying?
* In combine worker functions, `it.resultChs` removes the channel when the sub channels are closed. I tried using index in initial coding, but the index is warngled because of this.

winnie-byun replied:
@ehnuje Could you take a look with the new code? I couldn't still found out a way to simplify the parameters.

ehnuje commented:
Just FYI, geth has changed `Iteratee` interface in April.
https://github.com/ethereum/go-ethereum/pull/20808

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+		}
+
+		// look for smallest key
+		minEntry, minIdx := entries[0], 0
```
How about using minheap for finding smallest key?

winnie-byun replied:
Review reflected

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+// Next gets the next item from iterators.
+// The first call of Next() might take 50ms.
+func (it *shardedDBIterator) Next() bool {
+	maxRetry := 5
```
Why is this retry logic necessary?

ehnuje replied:
How about using a const instead of declaring a variable every time?

KimKyungup replied:
I think `Next()` should return `true` if there is a next item. So we need to modify the implemnetation of `Next()`.
```
// Returns true if there is a next instruction and moves on.
```

winnie-byun replied:
I removed the retry logic. As you mentioned, I found out it doesn't need waiting. `Next()` always wait for the next item or the channel to be closed.
Do you still think the comment is needed?

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+	var err error
+	for i, iter := range it.iterators {
+		if iter.Error() != nil {
+			logger.Error("[shardedDBIterator] error from iterator",
```
Need to move `err = iter.Error()` before the logger.

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+	for len(it.resultChs) != 0 {
+		// check if done
+		select {
+		case <-it.ctx.Done():
```
Can this case receive the signal, when we close blockchain?

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+				it.key, it.value = e.key, e.val
+				return true
+			} else {
+				logger.Error("[shardedDBIterator] Next is called on closed channel")
```
It seems that this log is always printed when the channel is closed. 
But `close(it.resultCh)` is not error case.

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+				return false
+			}
+		default:
+			logger.Debug("[shardedDBIterator] no value is ready")
```
I think default is not necessary. `case e, ok := <-it.resultCh:` is enough.
What do you think?

ehnuje replied:
If the channel is empty, then what would be the value of `e` here? Is it nil?

winnie-byun replied:
If the channel is empty, `case e, ok := <-it.resultCh:` is not run.
`e, ok := <-it.resultCh` will wait until a value arrives or the channel is closed.

KimKyungup replied:
@winnie-byun If there is no item, we can close the channel.

winnie-byun replied:
You are right. Thanks for pointing it out. The review is reflected.

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+			err = iter.Error()
+		}
+	}
+	return err
```
```suggestion
	return nil
```

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+		if iter.Error() != nil {
+			logger.Error("[shardedDBIterator] error from iterator",
+				"err", err, "shardNum", i, "key", it.key, "val", it.value)
+			err = iter.Error()
```
```suggestion
			return iter.Error()
```

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+	for i, iter := range it.iterators {
+		if iter.Error() != nil {
+			logger.Error("[shardedDBIterator] error from iterator",
+				"err", err, "shardNum", i, "key", it.key, "val", it.value)
```
```suggestion
				"err", iter.Error(), "shardNum", i, "key", it.key, "val", it.value)
```

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+	key, val []byte
+}
+
+// shardedDBIterator iterates all items of each shardDB.
```
```suggestion
// shardedDBIteratorUnsorted iterates all items of each shardDB.
```

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+}
+
+// shardedDBIterator iterates all items of each shardDB.
+// This is useful when you want to get items in serial.
```
```suggestion
// This is useful when you want to quickly get items in parallel.
```

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+				} else {
+					it.resultCh <- e
+				}
+			default:
```
Does this default case make infinity loop?
I think we should not use default case. Without this default, sleeping/waiting the channel is better.

So how about making each goroutine that pass items from each `ch` to `it.resultCh`?

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+	return it
+}
+
+func (*shardedDBChanIterator) newChanWorker(it Iterator, resultCh chan entry, ctx context.Context) {
```
How about something like this?
```suggestion
func (*shardedDBChanIterator) chanWokerLoop(it Iterator, resultCh chan entry, ctx context.Context) {
```

winnie-byun replied:
I will take this suggestion. I also changed the combine worker function: `newCombineWorker` -> `combineWorkerLoop`

KimKyungup replied:
It was not changed???

winnie-byun replied:
It is changed now. Thank you. :)

KimKyungup commented on storage/database/sharded_database_test.go:
```diff
@@ -0,0 +1,271 @@
+
+// testIterator tests if given iterator iterates all entries
+func testIterator(t *testing.T, checkOrder bool, entriesFromIterator func(db shardedDB, entryNum int) []entry) {
+	entryNum := 500
```
Could you test several entryNum case like below?
- less than NumStateTrieShards
- same with NumStateTrieShards
- upper than NumStateTrieShards
```suggestion
	entryNum := 500
```

winnie-byun replied:
I changed `testIterator`. entryNum case is now
* shardedDBCombineChanSize + 10 
* NumStateTrieShards - 1
* NumStateTrieShards
* NumStateTrieShards + 1

winnie-byun commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,283 @@ func (pdb *shardedDB) Close() {
+}
+
+// Release stops all iterators, channels and workers
+// Even Release() is called, there could be
```
I finished the comment. Thanks you for letting me know :)

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,280 @@ func (pdb *shardedDB) Close() {
+				if !ok { // channel is closed
+					it.resultChs = append(it.resultChs[:i], it.resultChs[i+1:]...)
+					break chanIter
+				} else {
```
We don't need `else`.

winnie-byun replied:
Thanks for the review. However, the code is deleted. There is no combine process and sub worker puts item in one channel. 😢 🙏

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,280 @@ func (pdb *shardedDB) Close() {
+			case e, ok := <-ch:
+				if !ok { // channel is closed
+					it.resultChs = append(it.resultChs[:i], it.resultChs[i+1:]...)
+					break chanIter
```
We can remove `chanIter` tag.
```suggestion
					continue
```

winnie-byun replied:
Thanks for the review. However, the code is deleted. There is no combine process and sub worker puts item in one channel. 😢 🙏

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,280 @@ func (pdb *shardedDB) Close() {
+		e, ok := <-ch
+		if !ok { // channel is closed
+			it.resultChs = append(it.resultChs[:i], it.resultChs[i+1:]...)
+			entries = append(entries[:i], entries[i+1:]...)
+		} else {
+			entries[i] = e
+		}
```
```suggestion
			entries = append(entries[:i], entries[i+1:]...)
			continue
		} 
		
		entries[i] = e
```

winnie-byun replied:
Thanks for the review. However, the code is deleted while applying priority queue. 😢 🙏

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,280 @@ func (pdb *shardedDB) Close() {
+
+		// fill used entry with new entry
+		e, ok := <-it.resultChs[minIdx]
+		if !ok { // channel is closed
+			it.resultChs = append(it.resultChs[:minIdx], it.resultChs[minIdx+1:]...)
+			entries = append(entries[:minIdx], entries[minIdx+1:]...)
+		} else {
+			entries[minIdx] = e
+		}
```
```suggestion
		if !ok { // channel is closed
			it.resultChs = append(it.resultChs[:minIdx], it.resultChs[minIdx+1:]...)
			entries = append(entries[:minIdx], entries[minIdx+1:]...)
			continue
		} 
		
		entries[minIdx] = e
```

winnie-byun replied:
Thanks for the review. However, the code is deleted while applying priority queue. 😢 🙏

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,280 @@ func (pdb *shardedDB) Close() {
+		case e, ok := <-it.resultCh:
+			if ok {
+				it.key, it.value = e.key, e.val
+				return true
+			} else {
+				logger.Debug("[shardedDBIterator] Next is called on closed channel")
+				return false
+			}
```
```suggestion
				return true
			} 
			
			logger.Debug("[shardedDBIterator] Next is called on closed channel")
			return false
```

winnie-byun replied:
Thanks for the review. However, the code is deleted while removing the retry logic. 😢 🙏

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,280 @@ func (pdb *shardedDB) Close() {
+// Release stops all iterators, channels and workers
+// Even Release() is called, there could be some items left in the channel.
+func (it *shardedDBChanIterator) Release() {
+	it.cancel()
```
We need to `it.cancle()`?
If we cancel child iterator, they will finish itself.
```suggestion
	for i := 0; i < len(db.shards); i++ {
		it.iterators[i].cancle()
	}
```

KimKyungup commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,280 @@ func (pdb *shardedDB) Close() {
+func (*shardedDBChanIterator) newChanWorker(it Iterator, resultCh chan entry, ctx context.Context) {
+iter:
+	for it.Next() {
+		select {
+		case <-ctx.Done():
+			break iter
+		default:
+		}
```
if child DB iteratorcan get `cancle()`, this routine is not necessary?

winnie-byun replied:
I tried to use `Release()` and remove `context` and `cancel`. However, when I do, I need a lock, instead. A LevelDB iterator is not for concurrent access. A channel iterator keeps calling `Next()` and calling `Release()` on other go routine makes an `nil point error`. 
Using a lock is more complicated than a context. Moreover, context can also handle process exit and user's cancellation signal. @ehnuje

winnie-byun commented:
Dear reviewer,
I tried to apply your reviews and made a big change in codes and structure.
* https://github.com/klaytn/klaytn/pull/822/commits/4db66167c38d3a6396c949c50bc07a3731664b57 `shardedDBIterator.runCombineWorker` uses priority queue. To do so, made a priority queue with `[]byte`. (The code is mostly a copy of previous code. This is a temperory implementation. I will enhance the priority queue with another PR.)
* https://github.com/klaytn/klaytn/pull/822/commits/dd12ae66b6916b1d3799723ac5e37fdabe1d3345 `IteratorUnsorted`'s `combine worker` is deleted. `chanIterator` can put data in one channel depending on parameter.
* https://github.com/klaytn/klaytn/pull/822/commits/1a689027d0e556de2478deb705a0673ec51a92a2 Removed retry logic in `Next()`. In previous code, there was a waiting problem. However, the problem does not happen in current code, even if the channel size is small. The case is tested in `testIterator`.

I am still working on testcodes with various `entryNum`.
A new PR with changed `Iteratee` interface will be made. This will make the codes much simpler.

KimKyungup commented on common/prque/prque_byteSlice.go:
```diff
@@ -0,0 +1,65 @@
+// CookieJar - A contestant's algorithm toolbox
```
@winnie-byun Can we use golang native heap instead of 3rd party pkg?
http://golang.site/go/article/208-Go%EC%97%90%EC%84%9C-Heap-%EC%82%AC%EC%9A%A9%ED%95%98%EA%B8%B0

KimKyungup commented on common/prque/prque_byteSlice.go:
```diff
@@ -0,0 +1,65 @@
+import "container/heap"
+
+type PrqueByteSlice struct {
+	cont *sstackByte
```
@winnie-byun 
I think you don't need to use `sstackByte `.
Please check below.
http://golang.site/go/article/208-Go%EC%97%90%EC%84%9C-Heap-%EC%82%AC%EC%9A%A9%ED%95%98%EA%B8%B0

ehnuje commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,248 @@ func (pdb *shardedDB) Close() {
+// Each iterators fetch values in binary-alphabetical order.
+// This is useful when you want to operate on each items in parallel.
+type shardedDBChanIterator struct {
+	ctx    context.Context
+	cancel context.CancelFunc
```
When would be the situation we need timeout for iterator?
Can we just call `Release` from caller codes?

winnie-byun replied:
A context is not solely for timeout. It can also handle process exit and user's cancellation signal. It is similar with a signal channel. (We usually used `chan struct{}` type before.) It is different that even though a process is exited unexpectedly, it is guaranteed that the go routine will end.

ehnuje replied:
I asked since other iterators do not receive context to catch the signal to exit gracefully. If there are something needs to be done when it catches the signal, it should do that. But what it does here is just close the channel and terminates the related goroutines. If these are something needs to be done, don't we need to use context to every single goroutine to kill them gracefully? For example, database, it should be gracefully killed to ensure that the database has written all the things so that we can retrieve the data later if it is killed gracefully. However, I don't think this is the case for the iterators. How do you think?

ehnuje commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,248 @@ func (pdb *shardedDB) Close() {
+
+// runCombineWorker fetches any key/value from resultChs and put the data in resultCh
+// in binary-alphabetical order.
+func (it shardedDBIterator) runCombineWorker() {
```
Is it something intended?
```suggestion
func (it *shardedDBIterator) runCombineWorker() {
```

ehnuje reviewed (approved):
LGTM except for minor comments

winnie-byun commented on storage/database/sharded_database.go:
```diff
@@ -151,102 +154,248 @@ func (pdb *shardedDB) Close() {
+		copy(val, it.Value())
+		resultCh <- entry{key, val}
+	}
+	it.Release()
+	if sit.shardNum--; sit.combinedChan && sit.shardNum > 0 {
+		return
+	}
+	close(resultCh)
```
Could you explain why you think that way?
Each channel worker closes its own iterator. Therefore `it.Release()` should be called allways.
If the worker is using `combinedChan`, only the the last worker should close the channel. If not,  all workers always closes their channel, like an iterator release.

ehnuje replied:
I asked since running a worker serves both jobs, retrieving value from iterator and releasing the related resources. However, I understood your point. Thanks for your explanation.

jeongkyun-oh commented on storage/database/sharded_database_test.go:
```diff
@@ -0,0 +1,270 @@
+
+		// create sharded DB
+		db, err := newShardedDB(config, 0, config.NumStateTrieShards)
+		dbs[i] = *db
+		if err != nil {
+			t.Log("Error occured while creating DB")
+			t.FailNow()
+		}
```
Although it is a testcase, it is better to check the error first.
```suggestion
		if err != nil {
			t.Log("Error occured while creating DB")
			t.FailNow()
		}
		dbs[i] = *db
```

winnie-byun replied:
The logic is erased. Thank you for the comment.

jeongkyun-oh commented on storage/database/sharded_database_test.go:
```diff
@@ -0,0 +1,270 @@
+		}(config.Dir)
+
+		// create sharded DB
+		db, err := newShardedDB(config, 0, config.NumStateTrieShards)
```
How about using `MiscDB` instead of 0?

winnie-byun replied:
review reflected :)
